Implement `ReduceType` and start new `ocean.meta`

mihails-strasuns-sociomantic opened the pull request:
I have finally managed to fix my `ReduceType` implementation to not crash the compiler but trying to integrate it into existing ocean code has ended up with me starting new top-level package :)

There are two reason for that:

1) Purist one - this is something I was tempted to do long time ago because existing `ocean.core.Traits` in an unstructured kitchen sink containing many utilities that are not even traits. I propose more abstract `ocean.meta` name ("metaprogramming") with 3 distinct packages-categories - traits, type operations and code generation utils. This PR defines few building blocks for the former two.

2) Backwards compatibility concern - I wanted to keep resulting code as clean as possible and switch existing utilities to use new `ReduceType` as a backend but it is quite scary considering how much different convoluted code uses those utilities. Starting with independent package hierarchy allows to switch to new stuff on demand.

Time to destroy!

mihails-strasuns-sociomantic committed:
Start new ocean.meta package

Started by splitting qualifier related transition templates from
`ocean.transition`.

These utilities are heavily used even among most primitive of type traits -
separating them into own module with no imports help to simplify import chain
for all other modules as `ocean.transition` also defined more heavy-weight
stuff.

mihails-strasuns-sociomantic committed:
Define most basic set of type traits

The `Basic` module contains basic non-recursive type traits (mostly adapted from
old `ocean.core.Traits_tango` module) that are needed to distinguish between
different kinds of types.

mihails-strasuns-sociomantic committed:
Implement `ReduceType` type utility

Allows to reduce arbitrary type to a single value in a generic form by
recursively stripping its signature into smaller bits and processing with
user-supplied "reducer".

mihails-strasuns-sociomantic committed:
Implement some of indirection traits with reducer

Re-implements few of existing traits from `ocean.core.Traits` using new
`ReduceType` facility while keeping original unit tests to ensure new
implementation is sufficietly feature complete.

mihails-strasuns-sociomantic commented:
(misses release notes for now as I foresee some discussion)

gavin-norman-sociomantic commented:
`ocean.meta` is a good idea.

gavin-norman-sociomantic commented on src/ocean/meta/types/package_.d:
```diff
@@ -0,0 +1,11 @@
+    This package contains various template utilities that deduce types or values
+    by reflecting on supplied type arguments and doing some compile-time
+    manipulations on them.
+
```
Every module should have the copyright / license, right?

david-eckardt-sociomantic replied:
```d
import("licence.txt");
```

gavin-norman-sociomantic commented on src/ocean/meta/traits/Basic.d:
```diff
@@ -0,0 +1,424 @@
+
+*******************************************************************************/
+
+///
```
Spurious?

gavin-norman-sociomantic commented on src/ocean/meta/traits/Basic.d:
```diff
@@ -0,0 +1,424 @@
+
+/*******************************************************************************
+
+    Types are the types from which one or multiple other types cannot be
```
Odd wording: "Types are the types"

gavin-norman-sociomantic commented on src/ocean/meta/traits/Basic.d:
```diff
@@ -0,0 +1,424 @@
+
+*******************************************************************************/
+
+public enum ArrayKind
```
:+1: good idea.

gavin-norman-sociomantic commented on src/ocean/meta/traits/Basic.d:
```diff
@@ -0,0 +1,424 @@
+
+public enum ArrayKind
+{
+    NotArray    = 0,
```
Maybe note: `// defined as 0 so that the value of isArrayType!() can be used as a bool`.

gavin-norman-sociomantic commented on src/ocean/meta/traits/Basic.d:
```diff
@@ -0,0 +1,424 @@
+
+*******************************************************************************/
+
+template isFunctionType ( T )
```
`isCallableType`?

mihails-strasuns-sociomantic replied:
Nope, I have intentionally not included objects with `opCall` in this list. `isCallableType` would be a superset from `isFunctionType`.

gavin-norman-sociomantic commented on src/ocean/meta/traits/package_.d:
```diff
@@ -0,0 +1,11 @@
+    This package contains various template utilities that allow to reason about
+    types in some way. Most traits answer simple boolean question about
+    parameter type having certain qualities ("traits") thus the name.
+
```
License / copyright.

gavin-norman-sociomantic commented on src/ocean/meta/types/ReduceType.d:
```diff
@@ -0,0 +1,199 @@
+
+    Generic utility allowing to recursively visit an arbitrary type and reduce
+    its definition to some compile-time value. It is intended to be used as
+    implementation cornerstone for complex type traits to address recursive
```
A bit clearer: "...implementation cornerstone for complex type traits to avoid having to rewrite the recursive type reflection boilerplate wherever it's needed."

gavin-norman-sociomantic commented on src/ocean/meta/types/ReduceType.d:
```diff
@@ -0,0 +1,199 @@
+    though the base types (i.e. fields and referenced types):
+
+    ---
+    static struct ExampleReducer
```
Any reason not to do this as a documented unittest?

mihails-strasuns-sociomantic replied:
Yep, I can't have nested struct with templated method in D1.

mihails-strasuns-sociomantic replied:
(and declaring it outside of test block kills the documentation purpose)

gavin-norman-sociomantic replied:
Could be a D2 only unittest...?

mihails-strasuns-sociomantic replied:
There is already unittest with same code, I copied it here for the purpose of doc rendering only.

gavin-norman-sociomantic commented on src/ocean/meta/types/ReduceType.d:
```diff
@@ -0,0 +1,199 @@
+    );
+    static assert (
+        is(typeof(Reducer.accumulate)),
+        Reducer.stringof ~ " must define `accumluate` CTFE function"
```
"of type ..."

gavin-norman-sociomantic commented on src/ocean/meta/types/ReduceType.d:
```diff
@@ -0,0 +1,199 @@
+    );
+    static assert (
+        is(typeof(Reducer.visit!(int))),
+        Reducer.stringof ~ " must define `visit` templated function"
```
"of type ..."

gavin-norman-sociomantic commented on src/ocean/meta/traits/Indirections.d:
```diff
@@ -0,0 +1,172 @@
+        return (isArrayType!(T) == ArrayKind.Dynamic)
+            || (isArrayType!(T) == ArrayKind.Associative)
+            || isPointerType!(T)
+            || is(T == class)
```
How about interfaces?

david-eckardt-sociomantic replied:
`isReferenceType`?

gavin-norman-sociomantic commented on src/ocean/meta/traits/Indirections.d:
```diff
@@ -0,0 +1,172 @@
+
+*******************************************************************************/
+
+public template containsMultiDimensionalDynamicArrays ( T )
```
`has` or `contains`?

mihails-strasuns-sociomantic replied:
Original name used `has` but I think `contains` is more straightforward.

david-eckardt-sociomantic replied:
I like `contains`.

gavin-norman-sociomantic commented on src/ocean/meta/traits/Indirections.d:
```diff
@@ -0,0 +1,172 @@
+
+    Checks if T or any of its subtypes is a multi-dimensional dynamic array.
+
+    T and all of its subtypes, if any, are expected to be
```
What happens if these criteria are not met?

mihails-strasuns-sociomantic replied:
Was copied from old doc. With new implementation there are no restrictions on T at all ;)

gavin-norman-sociomantic reviewed (commented):
This looks excellent! Probably @david-eckardt-sociomantic will come up with hundreds of corner cases and spoil your fun ;)

david-eckardt-sociomantic commented on src/ocean/meta/types/Qualifiers.d:
```diff
@@ -0,0 +1,199 @@
+
+    Templates to define types with modified qualifiers based on some input
+    types.
+
```
Perhaps mention that this is D2 transition related? (Well, mostly... `Unqual` is needed in D2 as well.)

mihails-strasuns-sociomantic replied:
I added mention that some of templates are transition related but the module as a whole is not.

david-eckardt-sociomantic commented on src/ocean/meta/traits/Basic.d:
```diff
@@ -0,0 +1,424 @@
+
+public template isArrayType ( T )
+{
+    static if (is(T U == U[]))
```
`Unqual!(T)`!!! ****THIS IS THE CARDINAL ERROR I'VE BEEN HUNTING FOR WEEKS NOW!!!****

mihails-strasuns-sociomantic replied:
I don't see how it is applicable. Your case was about nested `static if` and re-assembling original type. Here I don't care at all if `T` is `const` or not - only that it matches `U[]` pattern (with `U` not being used).

david-eckardt-sociomantic replied:
See https://issues.dlang.org/show_bug.cgi?id=17353.

mihails-strasuns-sociomantic replied:
So, your point is that eventually this issue will be "fixed"? I rather doubt so, if anything is to happen, stripping of top level const will be simply documented as it is now.

In the meanwhile, with existing semantics the usage is correct.

david-eckardt-sociomantic replied:
For the record:
http://dlang.org/spec/expression.html#IsExpression
> `is ( Type Identifier == TypeSpecialization )`
>  The condition is satisfied if `Type` is semantically correct and is the same as `TypeSpecialization`. The Identifier is declared to be either an alias of the `TypeSpecialization` or, if `TypeSpecialization` is dependent on `Identifier`, the deduced type. 

With `T == const(U[])` the specialisation is `const(U)[]`, which is _not the same_ as `T` so the `is` should
yield false. Still it yields true, which is a contradiction to the D2 specification, so it is a compiler bug hiding a missing `Unqual`.

david-eckardt-sociomantic replied:
My point is that relying on a compiler bug may potentially be a suboptimal design decision. Do you prefer code that relies on compiler bugs for convenience?
Maybe this bug will never be fixed, maybe it will. Maybe it will declared to be a feature as you predict, but that would at least cause more `is` expression inconsistencies with type qualifiers. A perfectly consistent solution would be to just ignore all qualifiers in all `is(A == B)` and `is(A == B C)` expressions, which would make me very happy.

mihails-strasuns-sociomantic replied:
Depends on kind of the bug. If bug is unlikely to be fixed and workaround complicates code in many places, I'd totally consider relying on it like I did here. Rationale is that if it is going to be fixed, such intrusive change would require deprecations anyway and we will be able to adjust. If it is not going to be fixed, we pollute the code for no reason.

david-eckardt-sociomantic replied:
I disagree, but at least add a comment here. Some day someone with a `const` mindset will maintain, extend or refactor this code and have a hard time understanding how it works with a missing `Unqual` -- or add `static assert` consistency checks which mysteriously fail etc.

gavin-norman-sociomantic replied:
I agree with David. At very least, the bug should be mentioned here with an explanation of why the code currently works.

david-eckardt-sociomantic commented on src/ocean/meta/traits/Basic.d:
```diff
@@ -0,0 +1,424 @@
+
+template isPointerType ( T )
+{
+    static if (is(T U == U*))
```
`Unqual!(T)`

mihails-strasuns-sociomantic replied:
Same comment as for array.

david-eckardt-sociomantic commented on src/ocean/meta/traits/Basic.d:
```diff
@@ -0,0 +1,424 @@
+
+template isAggregateType ( T )
+{
+    const isAggregateType = is(T == class) || is(T == struct) || is(T == union);
```
`interface`?

david-eckardt-sociomantic commented on src/ocean/meta/types/ReduceType.d:
```diff
@@ -0,0 +1,199 @@
+        }
+        else static if (isArrayType!(T))
+        {
+            static if (is(T U : U[]))
```
Suggest using the return value of the nice new `isArrayType` template which already tells you what kind of array it is.

david-eckardt-sociomantic commented on src/ocean/meta/traits/Basic.d:
```diff
@@ -0,0 +1,424 @@
+{
+    static if (is(T U == U[]))
+        const isArrayType = ArrayKind.Dynamic;
+    else static if (is(T U : U[]))
```
What do you think about using `__traits(isStaticArray, T)` and `__traits(isAssociativeArray, T)` in D2 here?

david-eckardt-sociomantic commented on src/ocean/meta/types/ReduceType.d:
```diff
@@ -0,0 +1,199 @@
+        }
+        else static if (isPointerType!(T))
+        {
+            accumulate(result, reduce!(typeof(*T.init))());
```
- Suggest `is(Unqual!(T) U == U*)` since you need the pointer base type `U` for `reduce`.
- `isPointerType!(T)` (as well as `is(Unqual!(T) U == U*)`) yields `true` if `T` is a function pointer so the ` static if (isFunctionType!(T))` should be prior to this branch.

mihails-strasuns-sociomantic replied:
Any reasons to not use `typeof(*T.init)`? I find it both more reliable and readable than pattern matching.

david-eckardt-sociomantic commented on src/ocean/meta/traits/Indirections.d:
```diff
@@ -0,0 +1,172 @@
+
+    static Result visit ( T ) ( )
+    {
+        static if (is(T U == U[]))
```
`Unqual!(T)`

david-eckardt-sociomantic commented on src/ocean/meta/traits/Indirections.d:
```diff
@@ -0,0 +1,172 @@
+        float[][3][] y;
+        char[] z;
+    }
+
```
What about qualified array/element types? Should they be supported or not?

mihails-strasuns-sociomantic replied:
Totally should but I haven't added any tests of my own so far, only copied old ones. Will try adding some new

david-eckardt-sociomantic commented:
Pretty cool! Apart from type qualifiers no unhandled corner cares spring to my mind right now (however, experience shows we may encounter them when trying to use it in swarm neo :trollface:).

david-eckardt-sociomantic commented on src/ocean/meta/traits/Basic.d:
```diff
@@ -0,0 +1,424 @@
+
+    static assert (isFunctionType!(typeof(&foo1)));
+    static assert (isFunctionType!(typeof(foo2)));
+    static assert (isFunctionType!(typeof(&foo3)));
```
Looks like `typeof(foo3)` should be tested, too.

don-clugston-sociomantic commented on src/ocean/meta/traits/Basic.d:
```diff
@@ -0,0 +1,424 @@
+        T = type to check
+
+    Returns:
+        `true` if `T` is a signed integer type
```
The docs for all of these traits should say something like `built-in signed integer type`, so that it's clear that they don't include synthetic types like `BigInt`, `CustomFloat`.

mihails-strasuns-sociomantic committed:
Allow default seed/accumulate for ReduceType

ReduceType is primarily intended as a way to implement complex type traits
which answer simple true/false question. All such traits will use same
accumulator implementation that applies logical OR to all individual results,
thus it is practical to make it default.

mihails-strasuns-sociomantic committed:
Release notes for ocean.meta

mihails-strasuns-sociomantic commented:
Major update:

- addressed review comments
- enhanced some tests
- added `ocean.meta.traits.Basic.isCallableType` and `ocean.meta.traits.Basic.isReferenceType`
- added `ocean.meta.types.Arrays.ElementTypeOf`
- tweaked `ReduceType` implementation to allow non-static `visit` and accumulate - makes stateful reducers possible if needed
- tweaked `ReduceType` so that if `accumulate`/`seed` are not present, it tries to default to logical OR based one
- release notes

mihails-strasuns-sociomantic commented:
Have stumbled upon one major limitation with this approach - it dies with any type with recursive definition, i.e. `struct S { S* ptr; }`. Any ideas?

mihails-strasuns-sociomantic commented:
Adding simple cache fixed infinite recursion but is inherently unreliable as we can't get qualified names of types in D1:

```diff
@@ -145,6 +145,9 @@ private struct ReduceTypeImpl ( Reducer )
     // internal state
     Reducer reducer;
 
+    // To avoid infinite CTFE recursion with recursive types
+    Reducer.Result[string] cache;
+
     // Helper to calculate new value and update accumulator in one step
     private void accumulate ( T ) ( ref T accum, T next )
     {
@@ -178,7 +181,13 @@ private struct ReduceTypeImpl ( Reducer )
             auto result = false;
         }
 
-        accumulate(result, reducer.visit!(T)());
+        auto cached = T.stringof in cache;
+        if (cached)
+            return *cached;
+
+        auto value = reducer.visit!(T)();
+        cache[T.stringof] = value;
+        accumulate(result, value);
 
         static if (isPrimitiveType!(T))
         {
```

gavin-norman-sociomantic commented:
As verbally discussed, the limitation relating to recursive definitions should be mentioned in the release notes and doc. We can enhance this in the future.

david-eckardt-sociomantic commented on src/ocean/meta/traits/Basic.d:
```diff
@@ -0,0 +1,508 @@
+
+    struct Bacon { }
+
+    static assert (!isPointerType!(Bacon));
```
Python style?

mihails-strasuns-sociomantic replied:
What is a python style?

david-eckardt-sociomantic replied:
IIRC using `ham, eggs, bacon` as example symbols is Python style.

mihails-strasuns-sociomantic replied:
This test block was copied from `ocean.core.Traits` blindly :)

david-eckardt-sociomantic commented on src/ocean/meta/traits/Basic.d:
```diff
@@ -0,0 +1,508 @@
+{
+    const bool isCallableType =
+           isFunctionType!(T)
+        || is(typeof(T.opCall));
```
Minor point/out of interest: Wouldn't it be better to check if `typeof(&T.init.opCall)` is a function or delegate? (I could imagine that could be a hassle, though.)

david-eckardt-sociomantic commented on src/ocean/meta/types/Arrays.d:
```diff
@@ -0,0 +1,70 @@
+        static if (is(T U : U[]))
+        {
+            alias U ElementTypeOf;
+        }
```
No need for
```d
isArrayType!(T) == ArrayKind.Dynamic || isArrayType!(T) == ArrayKind.Static
```
because
```d
is(T U : U[])
```
yields the same result.

mihails-strasuns-sociomantic replied:
You asked me to do exactly other way around for the sake of readability last time ;)

david-eckardt-sociomantic replied:
Oops, really?! Fair enough then...

david-eckardt-sociomantic commented on src/ocean/meta/types/ReduceType.d:
```diff
@@ -0,0 +1,208 @@
+        else
+        {
+            static assert (false,
+                "Unexpected type kind during recursive iteration" ~ T.stringof);
```
Append a colon and a space to the message string.

david-eckardt-sociomantic commented on src/ocean/meta/traits/Basic.d:
```diff
@@ -400,6 +400,43 @@ unittest
+{
+    const isReferenceType =
+           isPointerType!(T)
+        || isFunctionType!(T)
```
Only function *pointers* are reference types, not functions, so this would yield a false positive if a function type is passed. (I can't think of a reasonable use case for this, but it can at least hide a bug in the code using this template.)

(Personal note: Function types are the most obscure and unnecessary D types I know -- neither reference nor value types -- and should IMHO be removed from the language because while being useless they need special handling like this.)

david-eckardt-sociomantic replied:
Just thought about it: `isPointerType` will detect function pointers so `is(T == delegate)` should be sufficient here.

mihails-strasuns-sociomantic replied:
You are 100% right.

david-eckardt-sociomantic reviewed (changes requested):
Added a few more requests for changes; the following are older and still need to be addressed:

https://github.com/sociomantic-tsunami/ocean/pull/101/commits/b301965f87908a8ba537e3bc9f6a55fbd5ee84f7#r115250581

https://github.com/sociomantic-tsunami/ocean/pull/101/commits/b301965f87908a8ba537e3bc9f6a55fbd5ee84f7#r115250977

mihails-strasuns-sociomantic commented:
>  the following are older and still need to be addressed

Could you please copy comments here? GitHub refuses to link them :(

mihails-strasuns-sociomantic commented:
Updated with documentation and release mention for recursive type problem. Also some fixes for @david-eckardt-sociomantic :)

david-eckardt-sociomantic commented:
- `Unqual!(T)` request in `ocean.meta.traits.Indirections:116` https://github.com/sociomantic-tsunami/ocean/pull/101/commits/b9da44c8099422a24b4832f29133b0337c10e26f#r115250581
- `containsMultiDimensionalDynamicArrays`: Question whether to support and test qualified types in `ocean.meta.traits.Indirections:214` https://github.com/sociomantic-tsunami/ocean/pull/101/commits/b9da44c8099422a24b4832f29133b0337c10e26f#r115250977

mathias-lang-sociomantic commented on src/ocean/meta/traits/Basic.d:
```diff
@@ -0,0 +1,508 @@
+
+*******************************************************************************/
+
+public template isComplexType( T )
```
Do we need this (and `isImaginaryType`) given they're deprecated in D2 and not used anywhere ?

mihails-strasuns-sociomantic replied:
It is used via `isPrimitiveType` trait.

mathias-lang-sociomantic commented on src/ocean/meta/traits/Basic.d:
```diff
@@ -0,0 +1,508 @@
+        const isArrayType = ArrayKind.Static;
+    else static if (is(typeof(T.init.values[0])[typeof(T.init.keys[0])] ==
+            Unqual!(T)))
+        const isArrayType = ArrayKind.Associative;
```
This syntax is really ugly.
Can't we reuse the much nicer https://github.com/sociomantic-tsunami/ocean/blob/8f7d8925ae0d1890e404c27b077a6f9f356400eb/src/ocean/core/Traits.d#L1431-L1497 instead ?

mihails-strasuns-sociomantic replied:
The syntax you link can only be used as part of template declaration in D1 and not in `is` expression. I don't want to add define new templates just to make a single line that no one will ever look at prettier ;)

mathias-lang-sociomantic commented on src/ocean/meta/traits/Basic.d:
```diff
@@ -0,0 +1,508 @@
+           is(T == function)
+        || is(typeof(*T.init) == function)
+        || is(T == delegate);
+}
```
I guess we completely ignore `@property` for now ?

mihails-strasuns-sociomantic replied:
Yep, and I don't see why we would ever want to pretend it exists.

mihails-strasuns-sociomantic commented:
Updated:
* Address b9da44c#r115250581

b9da44c#r115250977 should be already addressed - I added quite some qualifiers to tests of `containsMultiDimensionalDynamicArrays` last time.

mathias-lang-sociomantic commented on src/ocean/meta/types/Arrays.d:
```diff
@@ -0,0 +1,70 @@
+    else static if (isArrayType!(T) == ArrayKind.Associative)
+    {
+        alias AAElementType!(typeof(T.init.keys[0]), typeof(T.init.values[0]))
+            ElementTypeOf;
```
Ditto about the ugly syntax
